src/ginn.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GINN 1"
.TH GINN 1 "2012-07-30" "" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
ginn \- Gesture to Keypress / Mouse Button injector
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
ginn [<specific configuration file>]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBginn\fR is a program to test support gestures in legacy applications without the
need to rewrite them.  This support is made by injecting keypresses or mouse
buttons after matching rules from wishes.xml file.
.PP
You can define the matching rules by expressing \fIwishes\fR. Each wish contains
the name of the gesture, number of fingers, the matching rules.
The keypress to inject is defined by \fIkey\fR tag, and the mouse button by
\&\fIbutton\fR, you can add also up to 4 modifiers keys like \fIControl_L\fR.
.SS "\s-1EXAMPLES\s0"
.IX Subsection "EXAMPLES"
Sample wishes file:
.PP
.Vb 10
\&    <ginn>
\&    <global>
\&      <wish gesture="Drag" fingers="2">
\&        <action name="action5" when="update">
\&          <trigger prop="delta y" min="20" max="80"/>
\&          <key>Down</key>
\&        </action>
\&      </wish>
\&    </global>
\&    <applications>
\&      <application name="inkscape">
\&        <wish gesture="Pinch" fingers="2">
\&          <action name="action1" when="update">
\&            <trigger prop="radius delta" min="\-200" max="\-40"/>
\&            <key>KP_Subtract</key>
\&          </action>
\&        </wish>
\&      </application>
\&    </applications>
\&    </ginn>
.Ve
.PP
Gesture type can be: Drag, Pinch, Rotate or Tap
The time of injection can be selected by using \fIupdate\fR to inject when the
gesture is received or \fIfinish\fR to inject when the gesture is finished.
The choice should be written inside the \fIwhen\fR in the action line.
        <action name=\*(L"action1\*(R" when=\*(L"update\*(R">
.PP
The Open Input Framework gesture stack provide gesture updates which can be very small to
reach the desired interval specified in the trigger.
To provide more flexibility a gesture accumulation is performed by default to
the successive property values' updates for a desired gesture until it reaches
the min-max intervall or the gesture finish, then, the accumlation buffer is
reinitialized.
If you want to disable this feature you can specify it explicitly:
        <trigger prop=\*(L"delta x\*(R" min=\*(L"\-80\*(R" max=\*(L"\-40\*(R" accumulate=\*(L"false\*(R"/>
.PP
The attributes used in trigger can be obtained by running geistest, you can 
also refer to the list below.
You can use more than one trigger, but you need to use them in the same order
as they appear in the list below.
.PP
.Vb 10
\&    Sub\-list of general attributes
\&     "device id"       = 13
\&     "timestamp"       = 180598198
\&     "root window id"  = 254
\&     "event window id" = 159825921
\&     "child window id" = 159825921
\&     "focus x"         = 0.000000
\&     "focus y"         = 0.000000
\&     "gesture name"    = "Drag"
\&     "touches"         = 1
\&      
\&    Sub\-list of attributes used with Drag gesture
\&     "delta x"    = \-8.933333 
\&     "delta y"    = 22.000000
\&     "velocity x" = \-0.638095
\&     "velocity y" = 1.571429
\&     "position x" = 634.533386
\&     "position y" = 482.777771
\&
\&    Sub\-list of attributes used with Rotate gesture
\&     "angle delta"      = \-0.358113 
\&     "angular velocity" = \-0.029843
\&     "angle"            = \-1.159999
\&     
\&    Sub\-list of attributes used with Pinch gesture
\&     "radius delta"    = \-44.935688
\&     "radial velocity" = \-3.209692
\&     "radius"          = 180.466690
\&     
\&    Sub\-list of attributes used with Tap gesture
\&     "tap time"   = 74.000000 
\&     "position x" = 740.933350
\&     "position y" = 469.333344
\&     
\&    Last attribues in the general list
\&     "boundingbox x1" = 634.533386
\&     "boundingbox y1" = 482.777771
\&     "boundingbox x2" = 634.533386
\&     "boundingbox y2" = 482.777771
\&     "touch 0 id"     = 1079.000000
\&     "touch 0 x"      = 634.533386
\&     "touch 0 y"      = 482.777771
.Ve
.PP
The names of keys list can be obtained after removing the \s-1XK_\s0 prefix from
/usr/include/X11/keysymdef.h file
.PP
.Vb 9
\& The mouse buttons are written with their corresponding number:
\&    1   Left Button
\&    2   Middle Button
\&    3   Right Button
\&    4   Wheel Up
\&    5   Wheel Down
\&    6   Horizontal Wheel Left
\&    7   Horizontal Wheel Right
\&    8   "Back" in browsers
.Ve
.SH "FILES"
.IX Header "FILES"
\&\fI/etc/ginn/wishes.xml\fR
User customized wishes file.
.PP
\&\fI/usr/share/ginn/wishes.xml\fR
Basic ginn wishes file installed for reference.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
geistest
.SH "AUTHOR"
.IX Header "AUTHOR"
Mohamed Ikbel Boulabiar <boulabiar@gmail.com>
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright 2010\-2011 by Canonical, Ltd.


ChangeLog
------------------------------------------------------------
revno: 92
tags: 0.2.6
committer: Chase Douglas <chase.douglas@canonical.com>
branch nick: ginn
timestamp: Mon 2012-07-30 13:29:57 -0700
message:
  Release version 0.2.6
------------------------------------------------------------
revno: 91
committer: Chase Douglas <chase.douglas@canonical.com>
branch nick: ginn
timestamp: Mon 2012-07-30 13:29:09 -0700
message:
  Update to build against renamed geis package
------------------------------------------------------------
revno: 90 [merge]
committer: Chase Douglas <chase.douglas@ubuntu.com>
branch nick: ginn
timestamp: Wed 2012-04-25 18:25:37 -0700
message:
  Don't autorun ginn for now because it interferes with Unity gestures
------------------------------------------------------------
revno: 89 [merge]
fixes bug: https://launchpad.net/bugs/740213
author: Josh Fee <joshfee@gmail.com>
committer: Stephen M. Webb <stephen.webb@canonical.com>
branch nick: ginn
timestamp: Fri 2011-08-12 11:49:54 -0400
message:
  Check the loaded wishes to see which gesture are used.
------------------------------------------------------------
revno: 88
fixes bug: https://launchpad.net/bugs/770277
committer: Chase Douglas <chase.douglas@ubuntu.com>
branch nick: ginn
timestamp: Fri 2011-05-20 15:34:26 -0400
message:
  Fix segfault if bamf returns empty name
  
  Patch is from Martin Luder. Thanks Martin!
------------------------------------------------------------
revno: 87
tags: 0.2.5
committer: Stephen M. Webb <stephen.webb@canonical.com>
branch nick: ginn
timestamp: Tue 2011-05-17 08:46:31 -0400
message:
  Bumped version to 0.2.5
------------------------------------------------------------
revno: 86 [merge]
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Thu 2011-04-28 10:18:08 +0200
message:
  Fix styling issues in the man file,
  Add an examples of using the accumulator and the firing time features.
------------------------------------------------------------
revno: 85 [merge]
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Tue 2011-03-22 13:53:12 +0100
message:
  Added Stephen merge for .desktop file to allow ginn to run
  once the session starts
------------------------------------------------------------
revno: 84 [merge]
committer: Chase Douglas <chase.douglas@ubuntu.com>
branch nick: ginn
timestamp: Wed 2011-03-16 12:44:48 -0400
message:
  Merged scroll and pointer fixes branch
  
  Includes an indent reformat too.
------------------------------------------------------------
revno: 83 [merge]
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Fri 2011-02-04 18:27:18 +0100
message:
  Merge Stephen patch to fix Ginn config file install place.
------------------------------------------------------------
revno: 82 [merge]
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Fri 2011-02-04 15:32:32 +0100
message:
  Merge Stephen license Fix,
  Version 0.2.4
------------------------------------------------------------
revno: 81
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Thu 2011-02-03 10:59:19 +0100
message:
  Fix Bug #702982
------------------------------------------------------------
revno: 80
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Thu 2011-02-03 09:48:38 +0100
message:
  Fix crash with no app name return
------------------------------------------------------------
revno: 79
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Thu 2011-02-03 09:47:05 +0100
message:
  Add more informations in README.
------------------------------------------------------------
revno: 78
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Wed 2011-01-26 17:25:33 +0100
message:
  Add function for mixed injection
------------------------------------------------------------
revno: 77
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Wed 2011-01-26 17:13:53 +0100
message:
  add support for injecting mouse and keys
------------------------------------------------------------
revno: 76
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Wed 2011-01-26 16:46:43 +0100
message:
  Add more informations to the man file
------------------------------------------------------------
revno: 75 [merge]
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Wed 2011-01-19 14:02:48 +0100
message:
  Merge Stephen's addition of more directories
------------------------------------------------------------
revno: 74
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Tue 2010-12-21 18:56:26 +0100
message:
  Fix Emacs crashes when there is no .desktop file returned by libbamf.
------------------------------------------------------------
revno: 73 [merge]
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Fri 2010-12-10 15:37:04 +0100
message:
  Solve Makefile issues by merging Stephen fixes.
------------------------------------------------------------
revno: 72
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Fri 2010-12-10 15:05:59 +0100
message:
  Add a man file.
------------------------------------------------------------
revno: 71
tags: 0.2.2
committer: Mohamed-ikbel Boulabiar <boulabiar@gmail.com>
branch nick: ginn
timestamp: Fri 2010-12-03 11:07:36 +0100
message:
  Fix version and man page generation.
------------------------------------------------------------

src/ginn.c
/**
 * Copyright 2010 Canonical Ltd.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3 of the License, or (at your option) any later
 * version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 51
 * Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 *
 * Authors:
 * 	Mohamed-Ikbel Boulabiar <boulabiar@gmail.com>
 * 	Henrik Rydberg <rydberg@bitmath.org>
 *	Stephen M. Webb <stephen.webb@canonical.com>
 *
 */
#include "config.h"
#include <geis/geis.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/select.h>
#include <sys/stat.h>
#include <unistd.h>
#include <X11/Xlib.h>
#include <X11/keysym.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h> 


att config_attr[25] = {[0 ... 24] = {.attrName = "",.val = 0,.valMax = 0}
};

wish w1 = {.config_attr = {[0 ... 24] =
                           {.attrName = "",.val = 0,.valMax = 0}
                           }
,
.key = "",
.next = NULL
};

wish *wp, *wpEnd;
apps *ap;

pthread_mutex_t mtx = PTHREAD_MUTEX_INITIALIZER;
int wait=0;

static int inside(float x, float a, float b)
{
    return ((x <= b) && (x >= a));
}

void initw(struct wish *wp)
{
    int i;
    wp->button = 0;
    wp->key = "";
    wp->next = NULL;
    for (i = 0; i < 4; i++)
        wp->modifiers[i] = "";
    for (i = 0; i < 25; i++) {
        wp->config_attr[i].attrName = "";
        wp->config_attr[i].val = 0;
        wp->config_attr[i].valMax = 0;
    }
}

void inita(struct apps *ap)
{
    ap->next = NULL;
    ap->wp = NULL;
    ap->appName = "";
}

static void clear_accum_attrs(att * attrs)
{
    int i = 0;
    while (strcmp(attrs[i].attrName, "") != 0) {
        if (attrs[i].accumulate)
            attrs[i].accumVal = 0;
        i++;
    }
}

static void update_wishes()
{
    char *activeApp;
    apps *tmpAp = ap;
    int diff = 0;

    activeApp = (char *) getCurrentApp();
    printf(" --ActiveApp %s\n", activeApp);
    if (activeApp)
        diff = strcmp(activeApp, ap->appName);

    while (diff && ap->next) {
        ap = ap->next;
        diff = strcmp(activeApp, ap->appName);
    }

    if (!diff)
        wpEnd->next = ap->wp;
    else
        wpEnd->next = NULL;

    ap = tmpAp;
}

static int doIwait(){
    return wait;
}

static void makeWait(){
    pthread_mutex_lock( &mtx );
    wait=1;
    pthread_mutex_unlock( &mtx );
    usleep(100000);
    pthread_mutex_lock( &mtx );
    wait=0;
    pthread_mutex_unlock( &mtx );
    pthread_exit(3);
}

static void
gesture_match(GeisGestureType gesture_type,
              GeisGestureId gesture_id,
              GeisSize attr_count, GeisGestureAttr * attrs, int state)
{
    
    struct wish *topw;
    topw = wp;
    update_wishes();
    
    pthread_mutex_lock( &mtx );
    int w=doIwait();
    pthread_mutex_unlock( &mtx );
    if(w){
        clear_accum_attrs(wp->config_attr);
        wp = topw;
        return;
    }
    
    while (wp && (0 != strcmp(wp->key, "") || wp->button)) {
        int valid = 1;
        GeisFloat maxVal=0;
        if (gesture_type == wp->config_attr[0].val
            && attrs[8].integer_val == wp->config_attr[1].val) {
            int attrsI = 9, cAttrI = 2;
            do {
                if (0 ==
                    strcmp(attrs[attrsI].name,
                           wp->config_attr[cAttrI].attrName)) {
                    GeisFloat flv = attrs[attrsI].float_val;
                    GeisFloat val = wp->config_attr[cAttrI].val;
                if(abs(val)>abs(maxVal)){
                    maxVal=flv;
                }
                    GeisFloat valMax = wp->config_attr[cAttrI].valMax;
                    printf("DEBUG -- comparing %s and %s : ",
                           attrs[attrsI].name,
                           wp->config_attr[cAttrI].attrName);
                    printf("%.2f %.2f %.2f is inside: ",
                           attrs[attrsI].float_val,
                           wp->config_attr[cAttrI].val,
                           wp->config_attr[cAttrI].valMax);
                    printf("%i \n",
                           inside(attrs[attrsI].float_val,
                                  wp->config_attr[cAttrI].val,
                                  wp->config_attr[cAttrI].valMax));
                    //if this is a finish motion, we only care about the maxVal
                    if(wp->when==GINN_FINISH){
                        printf("This is a finish gesture: %d\n", wp->when);
                        if(inside(maxVal, val, valMax)){
                           //then we say that the motion has taken palce
                           valid=1;
                        }else{
                            //otherwise the motion has not taken place, that's for sure
                           valid=0;
                        }
                    }else{
                        //but if this is not a finish motion
                        valid=valid&&inside(flv, val, valMax);
                    }
                    
                    attrsI++;
                    cAttrI++;
                } else{
                    attrsI++;
                }
            } while ((0 != strcmp(wp->config_attr[cAttrI].attrName, "")) && attrsI < 18);
            if (valid) {
                if ((0 != wp->button)
                    && (0 != strcmp(wp->key, ""))) {
                    injMixBtnKey(XStringToKeysym(wp->key),
                                 wp->button, wp->modifiers);
                    printf("MIX -- MIX");
                } else {
                    if (0 != wp->button)
                        injButton(wp->button, wp->modifiers);
                    if (0 != strcmp(wp->key, ""))
                        injKey(XStringToKeysym(wp->key), wp->modifiers);
                }
                clear_accum_attrs(wp->config_attr);
                if(wp->when==GINN_FINISH){
                    pthread_t braker;
                    pthread_create( &braker, NULL, &makeWait, NULL);
                }
            }
        }
        if(wp->when==GINN_FINISH){
            clear_accum_attrs(wp->config_attr);
        }
        wp = wp->next;
    }
    wp = topw;
}

static Window getRootWindow()
{
    Display *display = NULL;
    Window window = 0;

    display = XOpenDisplay(NULL);
    if (!display) {
        fprintf(stderr, "error opening X11 display.\n");
        exit(1);
    }

    window = DefaultRootWindow(display);

    XCloseDisplay(display);

    return window;
}

static void print_attr(GeisGestureAttr * attr)
{
    fprintf(stdout, "\tattr %s=", attr->name);
    switch (attr->type) {
    case GEIS_ATTR_TYPE_BOOLEAN:
        fprintf(stdout, "%s\n", attr->boolean_val ? "true" : "false");
        break;
    case GEIS_ATTR_TYPE_FLOAT:
        fprintf(stdout, "%f\n", attr->float_val);
        break;
    case GEIS_ATTR_TYPE_INTEGER:
        fprintf(stdout, "%d\n", attr->integer_val);
        break;
    case GEIS_ATTR_TYPE_STRING:
        fprintf(stdout, "\"%s\"\n", attr->string_val);
        break;
    default:
        fprintf(stdout, "<unknown>\n");
        break;
    }
}

static void
gesture_added(void *cookie,
              GeisGestureType gesture_type,
              GeisGestureId gesture_id,
              GeisSize attr_count, GeisGestureAttr * attrs)
{
    int i = 0;
    fprintf(stdout, "Gesture type %d added\n", gesture_type);
    for (i = 0; i < attr_count; ++i)
        print_attr(&attrs[i]);
}

static void
gesture_removed(void *cookie,
                GeisGestureType gesture_type,
                GeisGestureId gesture_id,
                GeisSize attr_count, GeisGestureAttr * attrs)
{
    int i = 0;
    fprintf(stdout, "Gesture type %d removed\n", gesture_type);
    for (i = 0; i < attr_count; ++i)
        print_attr(&attrs[i]);
}

static void
gesture_start(void *cookie,
              GeisGestureType gesture_type,
              GeisGestureId gesture_id,
              GeisSize attr_count, GeisGestureAttr * attrs)
{
    int i = 0;
    fprintf(stdout, "Gesture type %d started\n", gesture_type);
    for (i = 0; i < attr_count; ++i)
        print_attr(&attrs[i]);

    // In GEIS v1, we know that the focus coords are in attrs 5 and 6
    movePointer((int)attrs[5].float_val, (int)attrs[6].float_val);
}

static void
gesture_update(void *cookie,
               GeisGestureType gesture_type,
               GeisGestureId gesture_id,
               GeisSize attr_count, GeisGestureAttr * attrs)
{
    int i = 0;
    fprintf(stdout, "Gesture type %d updated\n", gesture_type);
    for (i = 0; i < attr_count; ++i)
        print_attr(&attrs[i]);
    gesture_match(gesture_type, gesture_id, attr_count, attrs,
                  GINN_UPDATE);
}

static void
gesture_finish(void *cookie,
               GeisGestureType gesture_type,
               GeisGestureId gesture_id,
               GeisSize attr_count, GeisGestureAttr * attrs)
{
    int i = 0;
    fprintf(stdout, "Gesture type %d finished\n", gesture_type);
    gesture_match(gesture_type, gesture_id, attr_count, attrs,
                  GINN_FINISH);
}

GeisGestureFuncs gesture_funcs = {
    gesture_added,
    gesture_removed,
    gesture_start,
    gesture_update,
    gesture_finish
};

/*
 * Searches for a default config file.
 *
 * Returns a pointer to a config file name (which must be freed) or NULL
 * if no default config file was found.
 */
static char *ginn_default_config()
{
    static const char default_file_name[] = "/wishes.xml";
    static const char *search_paths[] = {
        "etc",
        "../etc",
        ".",
        "$HOME/.ginn",
        GINN_CONFIG_DIR
    };
    static const int num_paths =
        sizeof(search_paths) / sizeof(const char *);
    int i;

    for (i = 0; i < num_paths; ++i) {
        struct stat sbuf;
        char *file_name = NULL;

        if (strstr(search_paths[i], "$HOME")) {
            char *home_dir = getenv("HOME");
            if (!home_dir) {
                continue;
            } else {
                char *cdr = index(search_paths[i], '/');
                size_t file_name_length = strlen(home_dir)
                    + strlen(cdr)
                    + strlen(default_file_name)
                    + 1;
                file_name = calloc(file_name_length, sizeof(char));
                strcpy(file_name, home_dir);
                strcat(file_name, cdr);
            }
        } else {
            size_t file_name_length = strlen(search_paths[i])
                + strlen(default_file_name)
                + 1;
            file_name = calloc(file_name_length, sizeof(char));
            strcpy(file_name, search_paths[i]);
        }
        strcat(file_name, default_file_name);
        int sres = stat(file_name, &sbuf);
        if (sres == 0) {
            fprintf(stdout, "Using wishes file %s\n", file_name);
            return file_name;
        }
        free(file_name);
    }

    return NULL;
}

int main(int argc, char *argv[])
{
    GeisStatus status = GEIS_UNKNOWN_ERROR;
    GeisXcbWinInfo xcb_win_info = {
        .display_name = NULL,
        .screenp = NULL,
        .window_id = getRootWindow()
    };
    GeisWinInfo win_info = {
        GEIS_XCB_FULL_WINDOW,
        &xcb_win_info
    };
    GeisInstance instance;
    struct ginn_config cfg;

    {
        char *config_file_name = NULL;
        if (argc < 2) {
            fprintf(stderr, "usage: %s <configxml>\n", argv[0]);
            config_file_name = ginn_default_config();
            if (config_file_name) {
                fprintf(stderr,
                        "using default configuration file %s ... \n",
                        config_file_name);
            }
        } else {
            config_file_name = strdup(argv[1]);
        }
        if (NULL == config_file_name) {
            fprintf(stderr, "Could not find Ginn wishes\n");
            return -1;
        }

        if (ginn_config_open(&cfg, config_file_name)) {
            fprintf(stderr, "Could not load Ginn wishes\n");
            return -1;
        }
        free(config_file_name);
    }

    ginn_config_print(&cfg);
    ap = (struct apps *) malloc(sizeof(struct apps));
    inita(ap);
    wp = (struct wish *) malloc(sizeof(struct wish));
    initw(wp);
    ginn_config_store(&cfg, wp, ap);
    wpEnd = wp;
    while (wpEnd->next)
        wpEnd = wpEnd->next;

    int pos = 0;
    printf("\n");
    while (strcmp(config_attr[pos].attrName, "")) {
        printf("DEBUG %d %s %.2f %.2f \n", pos,
               config_attr[pos].attrName, config_attr[pos].val,
               config_attr[pos].valMax);
        pos++;
    }
    
    /**
     * Check the loaded wishes to see which gestures are used
     * and add these gestures to a NULL terminated list "sub_gestures_list"
     * 
     * This list is used to subscribe only the necessary gestures
     * to GEIS. This will avoid issues with 2 finger scrolls and other gestues
     * being dropped when not accounted for in the wishes file.
    **/
    
    apps * appPtr = ap;
    wish * wishPtr;
    char * sub_gestures_list[17] = {NULL};
    int finished = 0;
    
    /**
     * Loop through each wish inside ap
     * After all of the wishes in ap are accounted for
     * add global wishes
    **/ 
    
    while(finished != 1)
    {
        // Check if we are finished with the application wishes
        if(appPtr == NULL)
        {
            wishPtr = wp;
            finished = 1;
        }else{
            wishPtr = appPtr->wp;
        }
        
        // Loop through each wish
        while(wishPtr)
        {
            int i;
            for(i = 0; i < 25; i ++)
            {
                
                char *sub_gesture = malloc(sizeof(char[16]));
                int pos = 0;
                
                // Check for Drag gesture
                if(strcmp(wishPtr->config_attr[i].attrName, "delta x") == 0 ||
                    strcmp(wishPtr->config_attr[i].attrName, "delta y") == 0 ||
                    strcmp(wishPtr->config_attr[i].attrName, "velocity x") == 0 ||
                    strcmp(wishPtr->config_attr[i].attrName, "velocity y") == 0
                    )            
                {
                    sprintf(sub_gesture, "Drag,touch=%i", (int)wishPtr->config_attr[1].val);
                }
                
                // Check for Rotate gesture
                else if(strcmp(wishPtr->config_attr[i].attrName, "angle delta") == 0 ||
                    strcmp(wishPtr->config_attr[i].attrName, "angular velocity") == 0 ||
                    strcmp(wishPtr->config_attr[i].attrName, "angle") == 0
                    )            
                {
                    sprintf(sub_gesture, "Rotate,touch=%i", (int)wishPtr->config_attr[1].val);
                }
                
                // Check for Pinch gesture
                else if(strcmp(wishPtr->config_attr[i].attrName, "radius delta") == 0 ||
                    strcmp(wishPtr->config_attr[i].attrName, "radial velocity") == 0 ||
                    strcmp(wishPtr->config_attr[i].attrName, "radius") == 0
                    )            
                {
                    sprintf(sub_gesture, "Pinch,touch=%i", (int)wishPtr->config_attr[1].val);
                }
                
                // Check for Tap gesture
                else if(strcmp(wishPtr->config_attr[i].attrName, "tap time") == 0)            
                {
                    sprintf(sub_gesture, "Tap,touch=%i", (int)wishPtr->config_attr[1].val);
                }
                
                // Deallocate sub_gesture if no gesture was found
                else
                {
                    free(sub_gesture);
                    sub_gesture = NULL;
                    continue;
                }
                
                /** Check the current gesture list
                 * Only add if the gesture is not already in the list
                 * Store in index of the first NULL found
                **/
                
                while(pos < 17)
                {
                    if(sub_gestures_list[pos] == NULL)
                    {
                        sub_gestures_list[pos] = sub_gesture;
                        break;
                    }else if(strcmp(sub_gestures_list[pos], sub_gesture) == 0)
                    {
                        break;
                    }
                    pos ++;
                }
                
                
            }
            // Iterate to the next wish
            wishPtr = wishPtr->next;
        }
        
        // If there are more applications, iterate to the next
        
        if(finished != 1)
        {
            appPtr = appPtr->next;
        }
    }
    
    // Show which gestures were subscribed
    printf("Gestures subscribed:\n");
    
    pos = 0;
    while(pos < 17)
    {
        if(sub_gestures_list[pos] != NULL)
        {
            printf("%s\n", sub_gestures_list[pos]);
        }
            
        pos ++;
    }
    
    // End sub gesture list creation 

    status = geis_init(&win_info, &instance);
    if (status != GEIS_STATUS_SUCCESS) {
        fprintf(stderr, "error in geis_init\n");
        return 1;
    }

    status = geis_configuration_supported(instance, GEIS_CONFIG_UNIX_FD);
    if (status != GEIS_STATUS_SUCCESS) {
        fprintf(stderr, "GEIS does not support Unix fd\n");
        return 1;
    }

    int fd = -1;
    status =
        geis_configuration_get_value(instance, GEIS_CONFIG_UNIX_FD, &fd);
    if (status != GEIS_STATUS_SUCCESS) {
        fprintf(stderr, "error retrieving GEIS fd\n");
        return 1;
    }

    status = geis_subscribe(instance, GEIS_ALL_INPUT_DEVICES, sub_gestures_list,        //  GEIS_ALL_GESTURES,
                            &gesture_funcs, NULL);
    if (status != GEIS_STATUS_SUCCESS) {
        fprintf(stderr, "error subscribing to gestures\n");
        return 1;
    }

    openDisplay();

    while (1) {
        fd_set read_fds;
        FD_ZERO(&read_fds);
        FD_SET(fd, &read_fds);
        int sstat = select(fd + 1, &read_fds, NULL, NULL, NULL);
        if (sstat < 0) {
            fprintf(stderr, "error %d in select(): %s\n", errno,
                    strerror(errno));
            break;
        }

        if (FD_ISSET(fd, &read_fds)) {
            geis_event_dispatch(instance);
        }
    }

    geis_finish(instance);
    closeDisplay();
    free(wp);
    return 0;
}
